k2/csrc/determinize.cc

```diff
@@ -5,423 +5,706 @@
+                          // state to this state (along whichever specific
+                          // sequence of symbols we took to get here)
+
+  // This constructor is for the start-state (state zero) of the input FSA.
+  MaxTracebackState(): state_id(0), arc_id(-1),
+                       prev_state(nullptr), forward_prob(0.0) { }
+
+  /**
+     @param [in] state_id  State in input FSA that this corresponds to
+     @param [in] src   Previous LogSumTracebackState that we'll point back
+                      to, or NULL
+     @param [in] incoming_arc_index  Arc-index in input FSA.
+                      Its src_state will equal src->state_id,
+                      its dest_state will equal state_id.
+     @param [in] arc_weight   Weight on the input arc
+   */
+  MaxTracebackState(int32_t state_id,
+                    const shared_ptr<MaxTracebackState> &src,
+                    int32_t incoming_arc_index,
+                    int32_t arc_weight):
+      state_id(state_id),
+      arc_id(incoming_arc_index),
+      prev_state(src),
+      forward_prob(element->forward_prob + arc_weight) { }
+
+  /*
+    This takes the same args as the constructor.  It will update the traceback
+    info if this incoming arc had higher weight.
+  */
+  void Accept(const shared_ptr<MaxTracebackState> &src,
+              int32_t arc_index, float arc_weight) {
+    double new_forward_prob = src->forward_prob + arc_weight;
+    if (new_forward_prob > forward_prob) {
+      forward_prob = new_forward_prob;
+      arc_id = arc_index;
+      prev_state = src;
+      // state_id doesn't change.
+    }
+  }
 };
 
+
+class LogSumTracebackState;
+
+
 /*
-  Conceptually a determinized state in weighted FSA determinization would
-  normally
-  be a weighted subset of states in the input FSA, with the weights normalized
-  somehow (e.g. subtracting the sum of the weights).
+  This struct is used inside LogSumTracebackState; it represents an
+  arc that traces back to a previous LogSumTracebackState.
+  A LogSumTracebackState represents a weighted colletion of paths
+  terminating in a specific state.
+*/
+struct LogSumTracebackLink {
 
-  Two determinized states are equal if the states and weights are the same.  To
-  ensure differentiability, our assumption is that in general no two arcs in the
-  input FSA have identical weights.  We argue that two determinized states can
-  always be represented as a base-state and a symbol sequence.  Imagine that we
-  follow arcs with that symbol sequence from the base-state, and then in case we
-  reach the same states in the different ways we always select the best path
-  from the base-state.  That process gives us a set of states and weights.  We
-  argue that this representation is unique.  (If not, it won't matter actually;
-  it will just give us an output that's less minimal than it could be).
+  shared_ptr<LogSumTracebackState> prev_state;
+                      // `prev_state` is the state that this points back to.
```

Comment:
`prev_state(src)` instead of `src(src)`, right?
https://github.com/danpovey/k2/blob/a79f527e32d6fe064f704e0987c263bbf7894e16/k2/csrc/determinize.cc#L237-L240